lib/canvas.js

```diff
@@ -186,61 +186,60 @@ Canvas.prototype.createSyncJPEGStream = function(options){
  * Return a data url. Pass a function for async support (required for "image/jpeg").
  *
  * @param {String} type, optional, one of "image/png" or "image/jpeg", defaults to "image/png"
- * @param {Object} opts, optional, options for jpeg compression (see documentation for Canvas#jpegStream)
+ * @param {Object|Number} encoderOptions, optional, options for jpeg compression (see documentation for Canvas#jpegStream) or the JPEG encoding quality from 0 to 1.
  * @param {Function} fn, optional, callback for asynchronous operation. Required for type "image/jpeg".
  * @return {String} data URL if synchronous (callback omitted)
  * @api public
  */
 
 Canvas.prototype.toDataURL = function(a1, a2, a3){
-  // valid arg patterns (args -> type, opts, fn):
+  // valid arg patterns (args -> [type, opts, fn]):
   // [] -> ['image/png', null, null]
+  // [qual] -> ['image/png', null, null]
+  // [undefined] -> ['image/png', null, null]
   // ['image/png'] -> ['image/png', null, null]
+  // ['image/png', qual] -> ['image/png', null, null]
   // [fn] -> ['image/png', null, fn]
   // [type, fn] -> [type, null, fn]
+  // [undefined, fn] -> ['image/png', null, fn]
+  // ['image/png', qual, fn] -> ['image/png', null, fn]
   // ['image/jpeg', fn] -> ['image/jpeg', null, fn]
   // ['image/jpeg', opts, fn] -> ['image/jpeg', opts, fn]
+  // ['image/jpeg', qual, fn] -> ['image/jpeg', {quality: qual}, fn]
+  // ['image/jpeg', undefined, fn] -> ['image/jpeg', null, fn]
 
-  var type;
+  if (this.width === 0 || this.height === 0) {
+    // Per spec, if the bitmap has no pixels, return this string:
+    return "data:,";
+  }
+
+  var type = 'image/png';
   var opts = {};
   var fn;
 
-  switch (arguments.length) {
-    case 0:
-      type = 'image/png';
-      break;
-    case 1:
-      if ('image/png' === a1) {
-        type = a1;
-      } else if ('function' === typeof a1) {
-        type = 'image/png';
-        fn = a1;
-      } else if ('image/jpeg' === a1) {
-        throw new Error('type "image/jpeg" only supports asynchronous operation');
-      } else {
-        throw new Error('invalid arguments');
-      }
-      break;
-    case 2:
-      if (('image/png' === a1 || 'image/jpeg' === a1) && 'function' === typeof a2) {
-        type = a1;
-        fn = a2;
-      } else if ('image/jpeg' === a1 && 'object' === typeof a2) {
-        throw new Error('type "image/jpeg" only supports asynchronous operation');
-      } else if ('image/png' === a1 && 'object' === typeof a2) {
-        throw new Error('type "image/png" does not accept an options object');
-      } else {
-        throw new Error('invalid arguments');
-      }
-      break;
-    case 3:
-      if ('image/jpeg' === a1 && 'object' === typeof a2 && 'function' === typeof a3) {
-        type = a1;
+  if ('function' === typeof a1) {
+    fn = a1;
+  } else {
+    if ('string' === typeof a1 && ('image/png' === a1.toLowerCase() ||
+      'image/jpeg' === a1.toLowerCase())) {
```

Comment:
Can we store the supported image formats in an array as a constant at the top of the file? `var FORMATS = ['image/png', 'image/jpeg']` and `FORMATS.contains(a1.toLowerCase())`?